(#5225) - Guard uses of console before calling it

schultyy opened the pull request:
PR for #5225 

This adds guards for all usages of the `console` object. It is checked if the console object itself is present and also the actual method to be called.

daleharvey commented on packages/pouchdb/tonic-example.js:
```diff
@@ -22,16 +22,21 @@ await db.put({
+if(typeof console !== undefined && 'log' in console) {
+  // later you can fetch your doc
+  console.log(await db.get('mydoc'))
+}
```
We dont need to do this in the example code

daleharvey commented:
This is a great start, thanks but i think we should clean the code up a little bit, if we created a new file `console.js` in https://github.com/pouchdb/pouchdb/tree/master/packages/pouchdb-utils/src and exported 2 functions `warn` and `error` that looked something like
```
function warn() { 
  if (typeof console !== undefined && 'warn' in console) {
    console.warn.apply(console, arguments);
  }
}
```
and used those functions it would be cleaner, what do you think? 

Thanks
Dale

daleharvey commented on packages/pouchdb-core/src/setup.js:
```diff
@@ -66,9 +66,11 @@ PouchDB.parseAdapter = function (name, opts) {
-          // log it, because this can be confusing during development
-          console.log('PouchDB is downgrading "' + name + '" to WebSQL to' +
-            ' avoid data loss, because it was already opened with WebSQL.');
+          if(typeof console !== undefined && 'log' in console) {
```
If we only expose `warn` and `error` in pouchdb-utils I think we should switch this to a warn

schultyy commented:
Sounds good. So everywhere where `console` is used now that new two functions are imported then?

schultyy commented:
What about the normal `console.log` statements? Do they stay as they are right now?

nolanlawson commented:
I would switch to using Dale's suggestion of `pouchdb-utils/src/console` as well. However, I would keep it simple and just make it one file, something like this:

```js
function console(type, msg) {
  if (typeof console !== 'undefined' && type in console) {
    console[type](message);
  }
}
```
```js
console('info', 'my info message');
```

nolanlawson commented on packages/pouchdb-adapter-idb/src/index.js:
```diff
@@ -883,7 +883,9 @@ function init(api, opts, callback) {
 
     idb.onabort = function (e) {
-      console.error('Database has a global failure', e.target.error);
+      if(typeof console !== undefined && 'error' in console) {
```
not sure if eslint enforces this, but we do `if (` not `if(`

schultyy commented:
@nolanlawson I implemented it as you suggested. There's this https://github.com/pouchdb/pouchdb/blob/master/packages/pouchdb-mapreduce/src/index.js#L33-L39 where I'm not sure if this should be kept as it is or also converted to the new `guardedConsole`. What do you think?

daleharvey commented:
That looks awesome thanks, since `pouchdb-mapreduce` already imports from `pouchdb-utils` I think its good to get all these consistent and make the switch there too, once thats done I think this is good to land, thanks :)

daleharvey commented:
Awesome, so one last issue, this is currently failing in CI, have you been able to install and build locally? as far as I can tell you need to 1, add pouchdb-utils to the dependencies in `packages/pouchdb/package.json` and 2. change the imports to `import { guardedConsole } .....` in the extras files to get this to build

daleharvey commented on packages/pouchdb-utils/src/guardedConsole.js:
```diff
@@ -0,0 +1,8 @@
+function guardedConsole(method) {
+  if (console !== 'undefined' && method in console) {
+    var args = Array.prototype.slice.call(arguments, 1);
+    console[method].apply(null, args);
```
This is failing in the browser, fairly sure it needs to be 
 
    console[method].apply(console, args);

daleharvey commented:
Heh sorry one more issue, posted the fix in comments

schultyy commented:
I will have a look in the evening and fix that :)

daleharvey commented:
ok awesome this is all good, I went and screwed up the coverage myself I think, but your code looks great, 👍 and merged

schultyy commented:
Great :)
